File: global search/SB2-Synthetic_Benchmark_2/OSE.m
function traj_list = OSE(max_time, time_step, model, c, lb, ub, omega, CR, max_iter, level, select_dims, input_dims)

    steps = max_time / time_step;
    timestamps = (0:steps-1)' * time_step;
    init_inputs = zeros(steps, input_dims);
    
    p = [];
    disp(timestamps)
    disp(init_inputs)
    u = [timestamps, init_inputs];
    T = max_time;
    
    [ta, init_outputs] = feval(model, p, u, T);
    traj_list = {{init_inputs, init_outputs}};

    for iter = 1:max_iter

        disp(["Iteration: ", num2str(iter)]);
        target_features = cell(1, level);

        for j = 1:level
            feature = zeros(length(select_dims), 4);
            for k = 1:length(select_dims)
                i = select_dims(k);
                t_idx = randi([1, steps]);
                dif = ub(t_idx, i) - lb(t_idx, i);
                new_f = (1 + 2 * c) * dif * rand() + lb(t_idx, i) - c * dif;
                feature(k, :) = [new_f, t_idx, i, dif];
            end
            target_features{j} = feature;
        end

        % Find the closest trajectory
        min_dist = Inf;
        closest_traj = {};
        for idx = 1:length(traj_list)
            traj = traj_list{idx};
            d = dist(traj{2}, target_features);
            if d < min_dist
                min_dist = d;
                closest_traj = traj;
            end
        end

        closest_inputs = closest_traj{1};
        new_inputs = closest_inputs;

        delta = omega * tan(pi * (rand(size(closest_inputs)) - 0.5));
        mask = rand(size(closest_inputs)) < CR;
        new_inputs(mask) = closest_inputs(mask) + delta(mask);
        
        u = [timestamps, new_inputs];

        [ta, new_outputs] = feval(model, p, u, T);
        %disp(new_outputs)
        traj_list{end+1} = {new_inputs, new_outputs};
    end
end

function d = dist(traj, features)
    % Flatten multi-dimensional cell array of features
    if iscell(features)
        features = cat(1, features{:});
    end

    target = features(:, 1);
    row_idx = features(:, 2);
    col_idx = features(:, 3);
    scale = features(:, 4);

    idx = sub2ind(size(traj), row_idx, col_idx);
    y2_vals = traj(idx);

    d = sum(((y2_vals - target) ./ scale) .^ 2);
end

File: global search/OSE.py
import matlab.engine

# Start MATLAB engine
eng = matlab.engine.start_matlab()

def OSE(init_state, max_time, time_step, model, c, lb, ub, omega, CR, max_iter=50, level=1, select_dims=[0], input_dims=1):
  steps = max_time / time_step
  init_inputs = np.zeros((int(steps),int(input_dims)))
  init_outputs = model(init_state, init_inputs, time_step)
  traj_list = [(init_inputs, init_outputs)]
  
  for _ in range(max_iter):
    print(_, len(traj_list))
    target_features = []
    for j in range(level):
      feature = []
      for i in select_dims:
        t = np.random.randint(0, steps)
        dif = (ub[t][i] - lb[t][i])
        new_f = (1 + 2 * c) * dif * np.random.rand() + lb[t][i] - c * dif
        feature.append((new_f, t, i, dif))
      feature = np.array(feature)
      target_features.append(feature)
    target_features = np.array(target_features)

    #print(target_features)
    #print('++++++++++++++++++++++++++++++++++++')
    closest_traj = min(traj_list, key=lambda traj: dist(traj[1], target_features))
    #print(closest_traj)
    # closest_inputs = closest_traj[0]
    # new_inputs = closest_inputs.deepcopy()
    # for i, u in enumerate(closest_inputs):
    #   delta = omega * np.tan(np.pi * (np.uniform(-1,1) - 0.5))
    #   if(np.random.rand() < CR):
    #     new_inputs[i] = u + delta
    #   else:
    #     new_inputs[i] = u
    closest_inputs = closest_traj[0]
    #print(closest_inputs)
    #print('++++++++++++++++++++++++++++++++++++')
    new_inputs = copy.deepcopy(closest_inputs)
    delta = omega * np.tan(np.pi * (np.random.uniform(-1, 1, size=closest_inputs.shape) - 0.5))
    mask = np.random.rand(*closest_inputs.shape) < CR
    new_inputs[mask] = closest_inputs[mask] + delta[mask]
    #print(new_inputs)
    #print('++++++++++++++++++++++++++++++++++++')
    new_outputs = model(init_state, new_inputs, time_step)
    traj_list.append((new_inputs, new_outputs))
  return traj_list
